### ISSUE TITLE
'utf-8' codec can't decode byte 0xc3

### ISSUE THREAD

**Issue by Chipsterjulien** [urwid/urwid#58]
error: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 12: unexpected end of data
file: raw_display.py
line 769: you use l.decode('utf-8') but you should use l.decode('utf-8', 'ignore') and solve many bugs like this

**Comment by and3rson** [comment]
Is this still reproducable?

**Comment by mwpow3ll** [comment]
@and3rson I can reproduce this issue pretty reliably on 1.3.x under Debian 9 with the following steps to reproduce:

- Set the locale to German `localectl set-locale LANG=de_de.utf8`
- Set the keymap to German `localectl set-keymap de`
- Reboot the machine `reboot`
- Log back in and run an app built with urwid that uses an Edit widget
- Enter 'ß' into the textbox, on US keyboard this is the '-' key just to the right of '0'
- Application crashes with the following: error: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 0: unexpected end of data

**Comment by wardi** [comment]
`c3` is the first byte of the UTF-8 encoding of 'ß'. When running under a UTF-8 encoding urwid should wait for the complete sequence from the terminal and not return the first part separately, but IIRC it waits 100ms or so by default.

Have you overridden any delay settings? Of course Urwid shouldn't be crashing even if you have, but it's hard to decide what to do when the terminal has only send you part of a UTF-8 code point.

**Comment by mwpow3ll** [comment]
@wardi no, there has been no modification of delay settings. just setting the locale and keymap. Its near instantaneous the crash.

**Comment by mwpow3ll** [comment]
@and3rson @wardi Can this issue be reopened? I am seeing two messages which might actually shed some light.

- UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 0: unexpected end of data
- UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 0: ordinal not in range (128)

**Comment by mwpow3ll** [comment]
@wardi @and3rson I'm not sure I understand why, but Urwid's util.py tries to setlocale category LC_ALL to '' and when that throws an exception it then tries to determine the encoding from the default category of LC_CTYPE which results in default 'ascii' and no support above 128 which will crash when using de_DE.UTF-8 locale and entering 'ß'.

Can someone explain why urwid isn't just getting the default locale?

https://github.com/urwid/urwid/blob/16469df346bc4b4b7a1035e06048bb6f5d9aa111/urwid/util.py#L46-L50

modifying https://github.com/urwid/urwid/blob/16469df346bc4b4b7a1035e06048bb6f5d9aa111/urwid/util.py#L50

to `return locale.getdefaultlocale()[1]` seems to properly resolve my above issues. Would this be suitable workaround?
